ffi/capi/src/utf.rs

```diff
@@ -0,0 +1,122 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use alloc::borrow::Cow;
+
+use writeable::{impl_display_with_writeable, LengthHint, Writeable};
+
+use core::fmt;
+
+/// Implements [`Writeable`] for [`&[u8]`] according to the [WHATWG Encoding Standard](
+/// https://encoding.spec.whatwg.org/#utf-8-decoder).
+#[derive(Debug)]
+#[allow(clippy::exhaustive_structs)] // newtype
+pub struct PotentiallyInvalidUtf8<'a>(pub &'a [u8]);
+
+impl Writeable for PotentiallyInvalidUtf8<'_> {
+    fn write_to<W: core::fmt::Write + ?Sized>(&self, sink: &mut W) -> fmt::Result {
+        let mut remaining = self.0;
+        loop {
+            match core::str::from_utf8(remaining) {
+                Ok(valid) => {
+                    return sink.write_str(valid);
+                }
+                Err(e) => {
+                    // SAFETY: By Utf8Error invariants
+                    let valid = unsafe {
+                        core::str::from_utf8_unchecked(remaining.get_unchecked(..e.valid_up_to()))
+                    };
+                    sink.write_str(valid)?;
+                    sink.write_char(char::REPLACEMENT_CHARACTER)?;
+                    let Some(error_len) = e.error_len() else {
+                        return Ok(()); // end of string
+                    };
+                    // SAFETY: By Utf8Error invariants
+                    remaining = unsafe { remaining.get_unchecked(e.valid_up_to() + error_len..) }
+                }
+            }
+        }
+    }
+
+    fn writeable_length_hint(&self) -> writeable::LengthHint {
+        // Lower bound is all valid UTF-8, upper bound is all bytes with the high bit, which become replacement characters.
+        LengthHint::between(self.0.len(), self.0.len() * 3)
+    }
+
+    fn write_to_string(&self) -> Cow<str> {
+        match core::str::from_utf8(self.0) {
+            Ok(valid) => Cow::Borrowed(valid),
+            Err(e) => {
+                // SAFETY: By Utf8Error invariants
+                let valid = unsafe {
+                    core::str::from_utf8_unchecked(self.0.get_unchecked(..e.valid_up_to()))
+                };
+
+                // Let's assume this is the only error
+                let mut out = alloc::string::String::with_capacity(
+                    self.0.len() + char::REPLACEMENT_CHARACTER.len_utf8()
+                        - e.error_len().unwrap_or(0),
+                );
+
+                out.push_str(valid);
+                out.push('�');
```

Comment:
Nit, optional: I would prefer explicitly writing out the code point so that I know this is a literal replacement character, U+FFFD, and not some other weird Unicode question mark symbol.

```suggestion
                out.push('\u{FFFD}');
```

I say this aware of the following section in the style guide.

https://github.com/unicode-org/icu4x/blob/main/documents/process/style_guide.md#render-visible-characters-in-docs-and-code--suggested

Reply:
No, I recognise the question mark as the replacement character, I don't recognise the escape sequence.